wgsl/index.bs

```diff
@@ -4557,11 +4555,271 @@ TODO: *Stub*: Expression evaluation
+* Look at which vertices are reachable from "CF_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+    * (optional, makes the analysis faster) Remove from the graph all vertices that have been visited
+* Look at which vertices are reachable from "Value_return"
+    * Remember for later whether this set includes "False" and/or any of the "arg_i"
+
+Note: Anything which is not explicitly remembered, including the entire graph can be destroyed at this point. Also note that the information that we need to remember to analyze the caller is linear in the number of parameters of the function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
+
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+We achieve this by assigning to each statement a set of *behaviors* out of {Return, Break, Continue, Discard, Nothing}.
+
+TODO: do we need an extra behavior for possibly infinite loops? I should check whether some form of reconvergence happens after something like `while(tid mod 2) {}`.
+
+If the set of behaviors for a statement is a singleton, we take that to mean that reconvergence happened, and we pick the vertex corresponding to control-flow at the start of the statement as the vertex corresponding to control-flow at the exit of the statement, overriding the other rules below.
 
-### Divergence and reconvergence TODO ### {#divergence-reconvergence}
+The rules for analyzing statements take as argument both the statement itself and the vertex corresponding to control-flow at the beginning of it (which we'll note "CF" below) and return the following:
 
-### Uniformity restrictions TODO ### {#uniformity-restrictions}
+* A set of behaviors for the statement
+* A vertex corresponding to control-flow at the exit of it
+* A set of new vertices and edges to add to the graph
+
+In the table below, `CF1, S -> CF2, B` means "run the analysis on S starting with control-flow CF1, apply the required changes to the graph, and name the resulting control-flow CF2, and the resulting set of behaviors B".
+Similarly, `CF, E -> CF, V, B` means "run the analysis on expression E, starting with control-flow CF, apply the required changes to the graph, and name the resulting control-flow vertex CF, the resulting value vertex V, and the resulting set of behaviors B" (see next section for the analysis of expressions).
+
+We have a similar set of rules for expressions in left-value positions, that we denote by `CF, E => CF, L, B`. Instead of computing the vertex which corresponds to the uniformity of the value, it computes the vertex which corresponds to the uniformity of the memory location we are adressing.
+
+<table class='data'>
+  <caption>Uniformity rules for statements</caption>
+  <thead>
+    <tr><th>Statement<th>New vertices<th>Precondition<th>Behaviors<th>Resulting control-flow vertex<th>New edges
```

Comment:
`for` loops don't seem to be listed here. We should either
A) Describe the desugaring in a more robust way in the "`for` statement" section of the spec, and link to it here
B) Add a row in the table for the `for` loop.

Reply:
I got confused by the "new vertices" column heading of the chart, because the "resulting control-flow vertex" is a new vertex, but doesn't appear in the "new vertices" column.

Reply:
I'm having trouble understanding the "precondition" column. Above, you say

> In the table below, `CF1, S -> CF2, B` means "run the analysis on S starting with control-flow CF1, apply the required changes to the graph, and name the resulting control-flow CF2, and the resulting set of behaviors B".

And these `CF1, S -> CF2, B` rules are inside the "precondition" column of the table. How can a precondition involve running an analysis and modifying a graph? I always thought a precondition was just a claim that either is true or untrue, and didn't modify state or anything.

Reply:
I did not list the rules for "for" loops, because they are supposed to desugar to "loop".

The "resulting control-flow vertex" may be a brand new vertex (in which case it will also be in the "new vertices column"), but it may instead be the result of the application of a rule on a sub-expression.

You are right that "precondition" is a bit of an abuse of language here, I should not have kept it from more conventional type systems. Maybe "recursive analyses" or something like that is a better name for this column.